coding_standard_c.txt

```diff
@@ -0,0 +1,147 @@
+}
+
+while(!labeled_correctly(sample)) {
+     /* ... */
+}
+
+A function that is exported for use in multiple source files should be prefixed with the source file (or object module) name it was defined in (not declared... infinite declarations are allowed, but only one definition).  For example, the file list.c may contain the implementation of a dynamic list ADT including an exported method for creating a list instance and an internal (static) method for overflow checking.  The first function might be named “list_create”, and the second, “is_overflowed”.  A function used only within its definition source file should be declared static (as should all variables global to only one file's scope), have no prefix, and must not be declared in any header files.  A function used externally from its definition source file should be declared extern in the source's associated header file.
+
+Files
+Typically, every header file has the same base filename as an associated source file.  Each header file is bound by conditional compilation commands using an all caps and underscored variant of the complete filename.  Header files should never define functions or variables.
+
+/*
+ * File: basic_file.h
+ * Header file block comment
+ */
+#ifndef BASIC_FILE_H
+#define BASIC_FILE_H
+/* standard #includes */
+/* project #includes */
+/* type/constant definitions */
+/* extern variable declarations */
+/* extern function declarations */
+#endif /* BASIC_FILE_H */
+/*
+ * File: basic_file.c
+ * Source file block comment...
+ */
+/* standard #includes */
+/* project #includes */
+#include “basic_file.h”
+/* externalized variable definitions */
+/* static variables (locally global) */
+/* static function declarations */
+/* externalized function definitions */
+/* static function definitions */
+
+Header files should only #include what is necessary to allow a file that includes it to compile.  Associated source files will always #include the header of the same name, but should #include files whose resources are used within the source even if they are already included in that header.  This provides a complete context for readers of the source file... i.e., they don't have to search through headers to determine  where a resource came from.  For example, consider a header and source that use FILE pointers declared in stdio.h.  The header must #include <stdio.h> to ensure files that include it will be able to compile.  The corresponding source #includes the associated header and therefore receives the benefit of stdio.h automatically.  The source would be unnerving if the FILE type was used and stdio.h was not #included.
+
+Variable Scope and types
+Variables should be declared and scoped to the smallest scope of use.
+
+/* This should be avoided */
+void foo() {
+	unsigned i;
+	do_something();
+	for(i=0; i < 100; i++) {
+		do_more(i);
+	}
+}
+
+/* Proper variable scope */
+void foo() {
+
+	do_something();
+	unsigned i;
+	for(i=0; i < 100; i++) {
+		do_more(i);
+	}
+}
+
```

Comment:
Having it local to use helps understand why the variable exists, and it also promotes people to declare variables for use instead of using a variable over and over for different things (since it's readily available declared at function entry and adding to that list might be "hard".) Let the compiler worry about optimizing memory and register usage of the variables...